src/drivers/imu/mpu6000/MPU6000.cpp

```diff
@@ -0,0 +1,989 @@
+	_px4_gyro(_interface->get_device_id(), (_interface->external() ? ORB_PRIO_HIGH : ORB_PRIO_DEFAULT), rotation),
+	_sample_perf(perf_alloc(PC_ELAPSED, "mpu6k_read")),
+	_measure_interval(perf_alloc(PC_INTERVAL, "mpu6k_measure_interval")),
+	_bad_transfers(perf_alloc(PC_COUNT, "mpu6k_bad_trans")),
+	_bad_registers(perf_alloc(PC_COUNT, "mpu6k_bad_reg")),
+	_reset_retries(perf_alloc(PC_COUNT, "mpu6k_reset")),
+	_duplicates(perf_alloc(PC_COUNT, "mpu6k_duplicates"))
+{
+	switch (_device_type) {
+	default:
+	case MPU_DEVICE_TYPE_MPU6000:
+		_px4_accel.set_device_type(DRV_ACC_DEVTYPE_MPU6000);
+		_px4_gyro.set_device_type(DRV_GYR_DEVTYPE_MPU6000);
+		break;
+
+	case MPU_DEVICE_TYPE_ICM20602:
+		_px4_accel.set_device_type(DRV_ACC_DEVTYPE_ICM20602);
+		_px4_gyro.set_device_type(DRV_GYR_DEVTYPE_ICM20602);
+		break;
+
+	case MPU_DEVICE_TYPE_ICM20608:
+		_px4_accel.set_device_type(DRV_ACC_DEVTYPE_ICM20608);
+		_px4_gyro.set_device_type(DRV_GYR_DEVTYPE_ICM20608);
+		break;
+
+	case MPU_DEVICE_TYPE_ICM20689:
+		_px4_accel.set_device_type(DRV_ACC_DEVTYPE_ICM20689);
+		_px4_gyro.set_device_type(DRV_GYR_DEVTYPE_ICM20689);
+		break;
+	}
+}
+
+MPU6000::~MPU6000()
+{
+	/* make sure we are truly inactive */
+	stop();
+
+	/* delete the perf counter */
+	perf_free(_sample_perf);
+	perf_free(_measure_interval);
+	perf_free(_bad_transfers);
+	perf_free(_bad_registers);
+	perf_free(_reset_retries);
+	perf_free(_duplicates);
+}
+
+int
+MPU6000::init()
+{
+
+#if defined(USE_I2C)
+	use_i2c(_interface->get_device_bus_type() == Device::DeviceBusType_I2C);
+#endif
+
+	/* probe again to get our settings that are based on the device type */
+	int ret = probe();
+
+	/* if probe failed, bail now */
+	if (ret != OK) {
+		PX4_DEBUG("CDev init failed");
+		return ret;
+	}
+
+	/* do init */
+	ret = CDev::init();
+
+	/* if init failed, bail now */
+	if (ret != OK) {
+		PX4_DEBUG("CDev init failed");
+		return ret;
+	}
+
+	if (reset() != OK) {
+		return ret;
+	}
+
+	return ret;
+}
+
+int MPU6000::reset()
+{
+	// if the mpu6000 is initialized after the l3gd20 and lsm303d
+	// then if we don't do an irqsave/irqrestore here the mpu6000
+	// frequently comes up in a bad state where all transfers
+	// come as zero
+	uint8_t tries = 5;
+	irqstate_t state;
+
+	while (--tries != 0) {
+		state = px4_enter_critical_section();
+
+		// Hold off sampling for 60 ms
+		_reset_wait = hrt_absolute_time() + 60000;
+
+		write_reg(MPUREG_PWR_MGMT_1, BIT_H_RESET);
+		up_udelay(10000);
+
+		// Wake up device and select GyroZ clock. Note that the
+		// MPU6000 starts up in sleep mode, and it can take some time
+		// for it to come out of sleep
+		write_checked_reg(MPUREG_PWR_MGMT_1, MPU_CLK_SEL_PLLGYROZ);
+		up_udelay(1000);
+
+		// Enable I2C bus or Disable I2C bus (recommended on data sheet)
+		write_checked_reg(MPUREG_USER_CTRL, is_i2c() ? 0 : BIT_I2C_IF_DIS);
+
+		px4_leave_critical_section(state);
```

Comment:
I am confused about why there are so many of these enter/exit critical sections.

Reply:
1. Are we doing the correct power management on boards other than fmu-v2/v3?